### PR TITLE
Adds subworkflow for A. baumannii, includes Kaptive task (K & O typing)

### PR DESCRIPTION
**Pull request by erikwolfsohn**
Added Kaptive task to detect Acinetobacter baumanii surface antigen synthesis loci

Will address issue [#99](https://github.com/theiagen/public_health_bacterial_genomics/issues/99)

Tested successfully in miniwdl & Terra

### CODE REVIEW

**Comment by erikwolfsohn** [comment]
Added Kaptive task to detect Acinetobacter baumanii surface antigen synthesis loci

Will address issue [#99](https://github.com/theiagen/public_health_bacterial_genomics/issues/99)

Tested successfully in miniwdl & Terra

**Comment by kapsakcj** [comment]
I'm working to resolve the pytest error over on my PR, once that is resolved I'll review your PR here. Thanks for the PR and appreciate your patience!

**Comment by erikwolfsohn** [comment]
> I'm working to resolve the pytest error over on my PR, once that is resolved I'll review your PR here. Thanks for the PR and appreciate your patience!

Thanks Curtis! I had a look this morning and I'm not sure what's causing the pytest error on my PR. It looks like mine it's failing here: `2022-08-23 00:08:31.659 wdl.w:theiaprok_illumina_pe.t:call-plasmidfinder.stderr VERBOSE ../command: line 13: plasmidfinder.py: command not found`

**Comment by kapsakcj** [comment]
Can you please sync your fork & branch with the changes I just made to `main` on the Theiagen repo? Hopefully this will resolve the Pytest errors. We think it was related to using a newer version of `miniwdl` for testing. So we downgraded to miniwdl `1.5.2` which allows things to pass without those odd issues.

Should be able to make a PR to your forked-repo to bring your branch `erikwolfsohn:kaptive-task` up-to-date with `Theiagen:main` (apart from the new Kaptive-related stuff) using this: https://github.com/erikwolfsohn/public_health_bacterial_genomics/compare/kaptive-task...theiagen:public_health_bacterial_genomics:main

**Comment by erikwolfsohn** [comment]
Sounds good! They should be synced up now.

**Comment by kapsakcj** [comment]
OK, automated tests have passed 🎉 

As soon as I'm able, myself or someone else on the team will do a deeper review of the code and test a few Acinetobacter baumanii's in Terra with this branch.

Should be able to dive into it either tomorrow or Friday

**Comment by kapsakcj** [comment]
I think the `.dockstore.yml` needs to be replaced since it looks to be deleted in this PR. It should be in sync with what we have on the `main` branch: https://github.com/theiagen/public_health_bacterial_genomics/blob/main/.dockstore.yml

**Comment by erikwolfsohn** [comment]
Whoops. Should be corrected now.

File: tasks/species_typing/task_kaptive.wdl [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    # Print and save date
+    date | tee DATE
+    # Print and save version
+    kaptive.py --version | tee VERSION 
+    # Run Kaptive on the input assembly with the --all flag and output with samplename prefix
+    kaptive.py \
+    ~{'--start_end_margin ' + start_end_margin} \
+    ~{'--min_gene_id ' + min_identity} \
+    ~{'--min_gene_cov ' + min_coverage} \
+    ~{'--low_gene_id ' + low_gene_id} \
```
Comment (kapsakcj) [inline tasks/species_typing/task_kaptive.wdl:31]: Any particular reason why `--low_gene_id` and `--start_end_margin` are used for K locus querying and not below for the OC locus?

Comment (erikwolfsohn) [reply tasks/species_typing/task_kaptive.wdl:31]: There is a reason, and that reason is that I forgot to put them there 😅. I'll add them in now. 

I do have all those values set to the defaults currently, so there shouldn't be any change in behavior.

**Comment by kapsakcj** [comment]
We could expand this task to accommodate `Klebsiella pneumoniae samples, but honestly I don't think it's worth it.

The Kleborate tool, which is already a part of TheiaProk, runs `kaptive` anyways and the K and O locus are included in the kleborate_output_file TSV. The K & O locus calls just aren't floated individually to the Terra table. Which in my opinion should be done

**Comment by erikwolfsohn** [comment]
Running Kaptive through Kleborate is definitely more convenient since Kleborate handles locating the loci databases and calling Kaptive however many times is needed. That said, I did notice that the Kaptive version called by Kleborate is very old - no idea if that's out of some necessity or if the docker image just needs to be updated.

**Comment by kapsakcj** [comment]
>That said, I did notice that the Kaptive version called by Kleborate is very old - no idea if that's out of some necessity or if the docker image just needs to be updated.

Ah! I didn't realize, thanks. Yes, we need to update the staphb docker image to kleborate 2.2.0. We currently have 2.1.0 available as a docker image, and are using kleborate 2.0.4 in the kleborate WDL task.

Kleborate 2.2.0 updates to Kaptive v2.0.0 (which I think is the most up-to-date for Klebsiella databases, though a bit behind for Acinetobacter)

I'll make a new docker image for kleborate 2.2.0 and we can update the default docker image in the WDL

**Comment by kapsakcj** [comment]
I was able to run this successfully in Terra on a few A baumannii's and upon a quick glance the outputs appear to be correct. (I have not tested the export_taxon_table functionality yet - usually save that for the end).

Really nice work putting this together! 👍 

I'm not ready to merge the PR yet, I want another Theiagen teammate to review and give feedback.

My high-level feedback is that we generally try to keep the number of output columns as low as possible, and only generate new outputs that are highly relevant & useful to typical users. 

For example, not every user will know what the `kaptive_k_other_outside_genes` column is and even if they do know, it may not be something they look at for every sample. This info is in the output TSV and my preference is to keep it there for advanced users that want to do the deep-dive on a sample. That way there are less output columns and only the more important & relevant output columns are in the Terra table.

To me (as someone who has not used this tool and is a bit naive) the important output columns are:

- kaptive_output_file_k - summary TSV file
- kaptive_output_file_oc - summary TSV file
- Best match locus - for both K and O - e.g. `KL7` and `OCL1`
- Best match type - for both K and O - e.g. `K7` and `OC1`
- Match confidence for both loci - e.g. `Very high`
- kaptive_version

Coverage, percent identity, and the other columns regarding inside/outside genes might be better kept in the summary TSVs instead of individual columns.

But I could be wrong here, if these are indeed important outputs that users should be aware of for A baumannii's then let's keep them.

**Comment by kapsakcj** [comment]
One suggestion for improvement - rename the output files to end in `.tsv` instead of `.txt`. Makes it easier to open the files with Excel afterwards.

Doesn't look like `kaptive` let's you name the file explicitly, so a `mv -v ~{samplename}_kaptive_out_oc_table.txt ~{samplename}_kaptive_out_oc_table.tsv` after the `kaptive` commands should do it. You'll also have to update the output block in the WDL task

**Comment by erikwolfsohn** [comment]
Morning Curtis. Thank you for doing the test runs in Terra!

> My high-level feedback is that we generally try to keep the number of output columns as low as possible, and only generate new outputs that are highly relevant & useful to typical users.

I agree, I think paring down the output to what you suggested is the best approach for now. As we get more involved w/SHARP, I think we'll have a better understanding of what data is of interest for the HAI projects we'll be participating in, and I can make changes then if needed.


> One suggestion for improvement - rename the output files to end in `.tsv` instead of `.txt`. Makes it easier to open the files with Excel afterwards.

Ah yeah, that is a good idea! Appreciate the feedback, I'll make these changes shortly.

**Comment by erikwolfsohn** [comment]
Just tested in Terra, everything looks good on my end.

**Comment by kapsakcj** [comment]
Thanks for making those changes, I think the outputs in the Terra table are much better this way.

I ran your updated workflow successfully in Terra, but I'm re-running right now to test export_taxon_table functionality. This can be tricky sometimes, so I want to make sure the resulting exported table has the same outputs as the source table

**Comment by kapsakcj** [comment]
Exported table looks identical 👍

**Review by kapsakcj — APPROVED** [review]
I'd like one other Theiagen person to review this, but all looks good on my end.

Thanks for the PR!

**Comment by kapsakcj** [comment]
@theiagen folks - for testing, import this workflow into your Terra workspace: https://dockstore.org/workflows/github.com/erikwolfsohn/public_health_bacterial_genomics/TheiaProk_Illumina_PE_Kaptive2:kaptive-task?tab=versions

**Comment by erikwolfsohn** [comment]
Glad to hear it. Thanks for the great feedback and taking the time to test it out!

**Comment by kapsakcj** [comment]
Agree with Emma^ I think that's an important distinction to make, between K locus and K type & likewise for O.

There are now a few conflicting files since we've merged other PRs recently - The theiagen folks can assist here, especially the `task_broad_terra_tools.wdl` which isn't easy to resolve. 

The other two files should be relatively straightforward to resolve conflicts

**Comment by erikwolfsohn** [comment]
Thank you both for the feedback! I'll make those changes and see if I can resolve the conflicts - if I have any trouble with the `task_broad_terra_tools.wdl` conflicts I'll reach out on Monday morning.

Have a great weekend!

**Comment by kapsakcj** [comment]
Just FYI I'm working on adding two new string outputs `kaptive_k_type` and `kaptive_oc_type`.

I'll also take care of the conflicts so we can get this PR tested again & merged soon

**Comment by kapsakcj** [comment]
Summary of the load of commits I just made:

- merged upstream changes to `theiagen:main` that were causing conflicts. It was changes from PR #141 Strep pneumo subworkflow
- Resolved conflict with `export_taxon_tables` task
- Kaptive task changes:
  - exposed `cpu` as an optional input param. Set to 4 by default. Added `kaptive.py -t ~{cpu}` to both kaptive.py commands.
  - lowered RAM to 8GB
  - added 2 new outputs `kaptive_k_type` and `kaptive_oc_type`
- added 2 new outputs ^ to theiaprok_illumina_pe workflow and merlin_magic workflow

I'm testing this now in Terra.

Also would like to add this to Theiaprok_illumina_SE once the testing on the PE workflow checks out ✔️

**Comment by kapsakcj** [comment]
OK. I think this PR is just about ready to merge into `main`!

CI should pass with my latest commit 2929b41

TheiaProk_Illumina_PE ran successfully in Terra and as expected, producing the 9 output columns from the Kaptive task. Export_taxon_tables functions as expected too (all columns get exported to new Terra data table)

TheiaProk_Illumina_SE is running right now in Terra, I'll come back and edit here with the status once it completes.

If you'd like to test the Single End workflow in Terra, you can import the workflow from Erik's dockstore account here: https://dockstore.org/workflows/github.com/erikwolfsohn/public_health_bacterial_genomics/TheiaProk_Illumina_SE:kaptive-task?tab=versions

@emmadoughty I believe we have addressed your requested change, as the `kaptive_k_type` and `kaptive_oc_type` are 2 new output columns that are parsed from the Kaptive output TSV. If the code looks good and you're satisfied with your testing, could you approve the PR?

**Comment by kapsakcj** [comment]
after discussion with @emmadoughty we'll remove the `kaptive_oc_type` output column from the workflows. It's been blank for all my test data and we don't think there is an established OC types.

We'll keep the output at the task level, for future applications and updates to `kaptive`, but will remove this extra/unnecessary output column

**Comment by kapsakcj** [comment]
Curtis todo's:
- [ ] rename output column `kaptive_oc_match` -> `kaptive_oc_locus`
- [ ] rename output column `kaptive_oc_confidence` -> `kaptive_ocl_confidence`
- [ ] rename output column `kaptive_k_match` -> `kaptive_k_locus`
- [ ] rename output column `kaptive_k_confidence` -> `kaptive_kl_confidence`

**Review by kapsakcj — APPROVED** [review]
Just approving once again since I've wrapped up my testing in Terra. Everything looks good-to-go in my opinion

**Review by emmadoughty — APPROVED** [review]

**Review by frankambrosio3 — APPROVED** [review]

**Pull request merged**